hedera-node/hedera-smart-contract-service-impl/src/main/java/com/hedera/node/app/service/contract/impl/exec/operations/CustomSelfDestructOperation.java

```diff
@@ -92,65 +97,135 @@ public OperationResult execute(@NonNull final MessageFrame frame, @NonNull final
             final var tbdAddress = frame.getRecipientAddress();
             final var proxyWorldUpdater = (ProxyWorldUpdater) frame.getWorldUpdater();
 
-            // Now proceed with the self-destruct
+            // In EVM the SELFDESTRUCT operation never fails.  It always sweeps ETH, and the contract
+            // is either deleted or not (per EIP-6780).
+            //
+            // In Hedera we have to allow for our semantics for transfers.  Notably, we can't
+            // transfer hbar unless the signature requirements are met on the transaction, we can't
+            // burn hbar, and we don't allow deletion of a contract which is a token treasury.
+            // There's also a restriction due to our performance guarantees: We don't allow contracts
+            // holding native tokens to self destruct because all the tokens would have to be
+            // transferred in the current `handleTransaction` call and if there were too many tokens
+            // it would be too expensive (in CPU/memory/database resources) to transfer them all.
+            //
+            // If the beneficiary account is the contract itself then we have two cases:
+            // * If (per EIP-6780) the contract is _not_ going to be deleted: That's ok.  SELFDESTRUCT
+            //   is a noop.  But if the contract _is_ going to be deleted and it has a balance of hbar
+            //   or any token then SELFDESTRUCT will fail.
+            // * Otherwise, if the beneficiary account is _not_ the contract itself then we fail the
+            //   SELFDESTRUCT if the contract owns any tokens.
+
+            final boolean contractCreatedInThisTransaction = frame.wasCreatedInTransaction(tbdAddress);
+            final boolean contractIsItsOwnBeneficiary = tbdAddress.equals(beneficiaryAddress);
+            final boolean contractIsToBeDeleted =
+                    switch (eip6780Semantics) {
+                        case NO -> true;
+                        case YES -> contractCreatedInThisTransaction;
+                    };
+
+            // inheritance == amount hbar to sweep
             final var inheritance =
                     requireNonNull(proxyWorldUpdater.get(tbdAddress)).getBalance();
             final var beneficiary = proxyWorldUpdater.get(beneficiaryAddress);
             final var beneficiaryIsWarm =
                     frame.warmUpAddress(beneficiaryAddress) || gasCalculator().isPrecompile(beneficiaryAddress);
-            final var cost = gasCalculator().selfDestructOperationGasCost(beneficiary, inheritance)
+            final var costWithoutBeneficiary = gasCalculator().selfDestructOperationGasCost(null, Wei.ZERO);
```

Comment:
Do we really want to charge no gas in these cases?  I see this was what we were doing before but we are doing _some_ work.

Reply:
The `Wei.ZERO` here isn't saying zero charge - it's saying the "inheritance" - i.e., the amount of the native coin to _sweep_ - is zero.  No balance.  This is because the gas charge (computed by the calculator, provided by Besu) is different if there is an inheritance as state must be changed: the address must be created if it has 0 balance, or otherwise it also depends on if the address is warm or cold.